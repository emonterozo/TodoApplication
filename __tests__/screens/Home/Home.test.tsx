import { fireEvent, screen, waitFor } from '../../test-utils';
import { renderMockGlobalContextProvider } from '../../context-utils';
import { Home } from '@app/screens';
import { mockedNavigate } from '../../../jest.setup';

describe('Home Screen', () => {
  it('should matches the snapshot', () => {
    renderMockGlobalContextProvider({
      children: <Home />,
    });
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('should navigate to Todo screen', async () => {
    renderMockGlobalContextProvider({
      children: <Home />,
    });

    const button = screen.getByTestId('add-todo');

    fireEvent(button, 'press');

    await waitFor(() => {
      expect(mockedNavigate).toHaveBeenCalledWith('Todo', {
        todo: undefined,
      });
    });
  });

  it('should filter todo if search input is not empty', async () => {
    renderMockGlobalContextProvider({
      children: <Home />,
      todo: [
        { id: '1234', title: 'First Todo', dueDate: '11/11/2024', status: 'Pending' },
        { id: '1235', title: 'Second Todo', dueDate: '11/11/2024', status: 'Completed' },
      ],
    });

    fireEvent.changeText(screen.getByTestId('search'), 'First Todo');

    await waitFor(() => {
      expect(screen.getByText('First Todo')).toBeTruthy();
      expect(screen.queryByText('Second Todo')).toBeNull();
    });
  });

  it('should filter todo if by status', async () => {
    renderMockGlobalContextProvider({
      children: <Home />,
      todo: [
        { id: '1234', title: 'First Todo', dueDate: '11/11/2024', status: 'Pending' },
        { id: '1235', title: 'Second Todo', dueDate: '11/11/2024', status: 'Completed' },
        { id: '1236', title: 'Third Todo', dueDate: '11/11/2024', status: 'In Progress' },
      ],
    });

    fireEvent(screen.getByTestId('selection-Pending'), 'press');

    await waitFor(() => {
      expect(screen.getByText('First Todo')).toBeTruthy();
      expect(screen.queryByText('Second Todo')).toBeNull();
    });

    // shall removed filter once same status is clicked
    fireEvent(screen.getByTestId('selection-Pending'), 'press');

    await waitFor(() => {
      expect(screen.getByText('First Todo')).toBeTruthy();
      expect(screen.queryByText('Second Todo')).toBeTruthy();
    });
  });
});

describe('Todo List', () => {
  beforeEach(() => {
    jest.useFakeTimers();
    jest.setSystemTime(new Date('2024-11-03'));
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('should render item if todo is not empty', () => {
    renderMockGlobalContextProvider({
      children: <Home />,
      todo: [{ id: '1234', title: 'Sample Todo', dueDate: '11/11/2024', status: 'Pending' }],
    });

    expect(screen.getByTestId('todo-1234').props.style.backgroundColor).toEqual('#fff');

    expect(screen.getByTestId('todo-1234')).toBeDefined();
  });

  it('should todo status background will be orange if status is not Completed', () => {
    renderMockGlobalContextProvider({
      children: <Home />,
      todo: [{ id: '1234', title: 'Sample Todo', dueDate: '11/11/2024', status: 'Pending' }],
    });

    const bgStyles = screen.getByTestId('todo-status-1234').props.style;

    expect(bgStyles[1].backgroundColor).toEqual('orange');
  });

  it('should todo status background will be green if status is Completed', () => {
    renderMockGlobalContextProvider({
      children: <Home />,
      todo: [{ id: '1234', title: 'Sample Todo', dueDate: '11/11/2024', status: 'Completed' }],
    });

    const bgStyles = screen.getByTestId('todo-status-1234').props.style;

    expect(bgStyles[1].backgroundColor).toEqual('green');
  });

  it('should navigate to Todo screen once todo item clicked', async () => {
    renderMockGlobalContextProvider({
      children: <Home />,
      todo: [{ id: '1234', title: 'Sample Todo', dueDate: '11/11/2024', status: 'Completed' }],
    });

    const button = screen.getByTestId('todo-1234');

    fireEvent(button, 'press');

    await waitFor(() => {
      expect(mockedNavigate).toHaveBeenCalledWith('Todo', {
        todo: { id: '1234', title: 'Sample Todo', dueDate: '11/11/2024', status: 'Completed' },
      });
    });
  });
});
