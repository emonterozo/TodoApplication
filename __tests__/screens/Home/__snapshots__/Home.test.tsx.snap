// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Screen should matches the snapshot 1`] = `
<RCTSafeAreaView>
  <TextInput
    onChangeText={[Function]}
    placeholder="Search title"
    style={
      {
        "borderWidth": 1,
        "height": 40,
        "marginBottom": 20,
        "marginHorizontal": 12,
        "padding": 10,
      }
    }
    testID="search"
    value=""
  />
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "borderRadius": 5,
            "marginHorizontal": 10,
            "padding": 15,
          },
          {
            "backgroundColor": "white",
          },
        ]
      }
      testID="selection-Pending"
    >
      <Text
        style={
          {
            "color": "black",
            "fontSize": 18,
          }
        }
      >
        Pending
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "borderRadius": 5,
            "marginHorizontal": 10,
            "padding": 15,
          },
          {
            "backgroundColor": "white",
          },
        ]
      }
      testID="selection-In Progress"
    >
      <Text
        style={
          {
            "color": "black",
            "fontSize": 18,
          }
        }
      >
        In Progress
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "borderRadius": 5,
            "marginHorizontal": 10,
            "padding": 15,
          },
          {
            "backgroundColor": "white",
          },
        ]
      }
      testID="selection-Completed"
    >
      <Text
        style={
          {
            "color": "black",
            "fontSize": 18,
          }
        }
      >
        Completed
      </Text>
    </View>
  </View>
  <RCTScrollView
    contentContainerStyle={
      {
        "padding": 16,
      }
    }
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View />
  </RCTScrollView>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 5,
          "marginHorizontal": 10,
          "padding": 15,
        },
        {
          "backgroundColor": "#EAB68F",
        },
      ]
    }
    testID="add-todo"
  >
    <Text
      style={
        {
          "color": "black",
          "fontSize": 20,
        }
      }
    >
      Add Todo
    </Text>
  </View>
</RCTSafeAreaView>
`;
